### PR TITLE
Decompile sound

### PR DESCRIPTION
**Pull request by ipatix**
This is the first attempt to fully decompile the ARM7 sound system.

### CODE REVIEW

**Review by ProjectRevoTPP — DISMISSED** [review]

File: arm7/lib/include/PM.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#ifndef GUARD_PM_H
+#define GUARD_PM_H
+
+#include "nitro/types.h"
+
+void PMi_SetControl(int ctrl);
```
Comment (red031000) [inline arm7/lib/include/PM.h:6]: does s32 or u32 work here?

File: arm7/lib/include/SND_capture.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#ifndef GUARD_SND_CAPTURE_H
+#define GUARD_SND_CAPTURE_H
+
+#include "nitro/types.h"
+
+void SND_SetupCapture(
```
Comment (red031000) [inline arm7/lib/include/SND_capture.h:6]: u32/s32?

File: arm7/lib/include/SND_channel.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#ifndef GUARD_SND_CHANNEL_H
+#define GUARD_SND_CHANNEL_H
+
+#include "nitro/types.h"
+
+void SND_SetupChannelPcm(int chnIdx,
```
Comment (red031000) [inline arm7/lib/include/SND_channel.h:6]: u32/s32?

File: arm7/lib/include/SND_exChannel.h [hunk 4]
```diff
@@ -1,8 +1,43 @@
...
+BOOL SND_StartExChannelPsg(struct SNDExChannel *chn, s32 duty, s32 length);
+BOOL SND_StartExChannelNoise(struct SNDExChannel *chn, s32 length);
+s32 SND_UpdateExChannelEnvelope(struct SNDExChannel *chn, BOOL step);
+void SND_SetExChannelAttack(struct SNDExChannel *chn, s32 attack);
+void SND_SetExChannelDecay(struct SNDExChannel *chn, s32 decay);
+void SND_SetExChannelSustain(struct SNDExChannel *chn, s32 sustain);
+void SND_SetExChannelRelease(struct SNDExChannel *chn, s32 release);
+void SND_ReleaseExChannel(struct SNDExChannel *chn);
+BOOL SND_IsExChannelActive(struct SNDExChannel *chn);
+struct SNDExChannel *SND_AllocExChannel(u32 channelMask,
```
Comment (red031000) [inline arm7/lib/include/SND_exChannel.h:23]: u32/s32?

File: arm7/lib/include/SND_global.h [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+#define GUARD_SND_GLOBAL_H
+
+#include "nitro/types.h"
+
+void SND_Enable(void);
+void SND_Disable(void);
+void SND_Shutdown(void);
+void SND_BeginSleep(void);
+void SND_EndSleep(void);
+void SND_SetMasterVolume(int vol);
```
Comment (red031000) [inline arm7/lib/include/SND_global.h:11]: u32/s32?

File: arm7/lib/include/SND_seq.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#define GUARD_SND_SEQ_H
+
+#include "nitro/types.h"
+
+#include "nitro/SND_bank_shared.h"
+
+void SND_SeqInit(void);
+void SND_SeqMain(BOOL update);
+
+void SND_StartSeq(int player, const void *seq, u32 offset, struct SNDBankData *bank);
```
Comment (red031000) [inline arm7/lib/include/SND_seq.h:11]: u32/s32?

File: arm7/lib/include/SND_util.h [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#ifndef GUARD_SND_UTIL_H
+#define GUARD_SND_UTIL_H
+
+#include "nitro/types.h"
+
+#define SND_VOL_DB_MIN (-723)
+
+u16 SND_CalcRandom(void);
+s8 SND_SinIdx(int index);
```
Comment (red031000) [inline arm7/lib/include/SND_util.h:9]: u32/s32?

File: arm7/lib/include/SND_work.h [hunk 8]
```diff
@@ -1,4 +1,13 @@
...
 #define GUARD_SND_WORK_H
 
-#endif //GUARD_SND_WORK_H
\ No newline at end of file
+#include "nitro/SND_work_shared.h"
+
+extern struct SNDWork SNDi_Work;
+extern struct SNDSharedWork *SNDi_SharedWork;
+
+void SND_SetPlayerLocalVariable(int player, int var, s16 value);
```
Comment (red031000) [inline arm7/lib/include/SND_work.h:9]: u32/s32?

File: arm7/lib/include/syscall.h [hunk 9]
```diff
@@ -2,5 +2,9 @@
 #define POKEDIAMOND_ARM7_SYSCALL_H
 
 void SVC_WaitByLoop(u32 ct);
+u8 SVC_GetVolumeTable(int idx);
```
Comment (red031000) [inline arm7/lib/include/syscall.h:5]: u32/s32?

File: arm7/lib/src/SND_capture.c [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+#include "registers.h"
+
+enum SNDLoop
+{
+    SND_CAP_LOOP = 0,
+    SND_CAP_ONESHOT = 1
+};
+
+void SND_SetupCapture(
+    int idx, int format, void *captureData, int size, BOOL loop, int capCtrlSrc, int capCtrlDst)
```
Comment (red031000) [inline arm7/lib/src/SND_capture.c:12]: u32/s32? or does it not match when replaced with that

File: arm7/lib/src/SND_channel.c [hunk 11]
```diff
@@ -0,0 +1,192 @@
+#include "SND_channel.h"
+
+#include "SND_work.h"
+#include "registers.h"
+
+static int sMasterPan = -1;
```
Comment (red031000) [inline arm7/lib/src/SND_channel.c:6]: u32/s32?

File: arm7/lib/src/SND_channel.c [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    vu32 *reg = &reg_SOUNDxCNT(idx);
+
+    u32 v = *reg;
+
+    // disable channel
+    v &= ~0x80000000;
+
+    // set hold flag
+    if (hold & 1)
+        v |= 0x8000;
```
Comment (red031000) [inline arm7/lib/src/SND_channel.c:93]: might it be an idea for this to be a constant, perhaps as an enum

File: arm7/lib/src/SND_channel.c [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+    if (sSurroundDecay > 0 && (1 << chnIdx) & 0xFFF5)
+    {
+        reg_SOUNDxCNT_VOL(chnIdx) = (u8)CalcSurroundDecay(sOrgVolume[chnIdx], pan);
+    }
+}
+
+BOOL SND_IsChannelActive(int chnIdx)
+{
+    return (reg_SOUNDxCNT_STAT(chnIdx) & 0x80) != 0;
```
Comment (red031000) [inline arm7/lib/src/SND_channel.c:135]: constant?

File: arm7/lib/src/SND_channel.c [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+void SNDi_SetSurroundDecay(int decay)
+{
+    sSurroundDecay = decay;
+
+    for (int i = 0; i < SND_CHANNEL_COUNT; i++)
+    {
+        // do not process channel 1+3 (capture playback channels)
+        if ((1 << i) & 0xFFF5)
```
Comment (red031000) [inline arm7/lib/src/SND_channel.c:170]: I'm seeing a lot of 0xFFF5, I'm not sure what this means, again, constant?

File: arm7/lib/src/SND_command.c [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+                    SetChannelTimer(cmd.arg[0], (int)cmd.arg[1]);
+                    break;
+                case SND_CMD_CHANNEL_VOLUME:
+                    SetChannelVolume(cmd.arg[0], (int)cmd.arg[1], (int)cmd.arg[2]);
+                    break;
+                case SND_CMD_CHANNEL_PAN:
+                    SetChannelPan(cmd.arg[0], (int)cmd.arg[1]);
+                    break;
+                case SND_CMD_SETUP_CHANNEL_PCM:
+                    SND_SetupChannelPcm((int)cmd.arg[0] & 0xFFFF,
```
Comment (red031000) [inline arm7/lib/src/SND_command.c:130]: a lot of numbers here, perhaps more constants?

File: arm7/lib/src/SND_command.c [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+    for (int i = 0; i < SND_CHANNEL_COUNT && channelMask != 0; i++, channelMask >>= 1)
+    {
+        if (channelMask & 1)
+            SND_SetChannelPan(i, pan);
+    }
+}
+
+static void StartTimer(u32 channelMask, u32 captureMask, u32 alarmMask, s32 unused)
+{
+    (void)unused;
```
Comment (red031000) [inline arm7/lib/src/SND_command.c:256]: ??
does the pragma not work here?

File: arm7/lib/src/SND_exChannel.c [hunk 17]
```diff
@@ -0,0 +1,685 @@
...
+        }
+
+        pan += chn->initPan;
+        if (chn->panRange != 127)
+        {
+            pan = (pan * chn->panRange + 0x40) >> 7;
+        }
+        pan += chn->userPan;
+
+        if (chn->envStatus == SND_ENV_RELEASE && vol <= -723)
```
Comment (red031000) [inline arm7/lib/src/SND_exChannel.c:189]: constant?

File: arm7/lib/src/SND_exChannel.c [hunk 18]
```diff
@@ -0,0 +1,685 @@
...
+            chn->volume = 0;
+            chn->flags.active = 0;
+        }
+        else
+        {
+            vol = SND_CalcChannelVolume(vol);
+            newTimer = SND_CalcTimer(chn->waveParam.timer, pitch);
+
+            if (chn->type == SND_CHN_TYPE_PSG)
+                newTimer &= 0xFFFC;
```
Comment (red031000) [inline arm7/lib/src/SND_exChannel.c:205]: constant?

File: arm7/lib/src/SND_exChannel.c [hunk 19]
```diff
@@ -0,0 +1,685 @@
...
+    }
+    else if (chn->id > 13)
+    {
+        return FALSE;
+    }
+    else
+    {
+        chn->type = SND_CHN_TYPE_PSG;
+        chn->dutyCycle = duty;
+        chn->waveParam.timer = 8006;
```
Comment (red031000) [inline arm7/lib/src/SND_exChannel.c:256]: constant?

File: arm7/lib/src/SND_exChannel.c [hunk 20]
```diff
@@ -0,0 +1,685 @@
...
+    if (chn)
+    {
+        chn->callback = NULL;
+        chn->callbackUserData = NULL;
+    }
+}
+
+void SND_StopUnlockedChannel(u32 channelMask, u32 weak)
+{
+    (void)weak;
```
Comment (red031000) [inline arm7/lib/src/SND_exChannel.c:409]: pragma?

File: arm7/lib/src/SND_global.c [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+    PMi_ResetControl(1);
+    reg_POWCNT2 &= ~1;
+}
+
+void SND_EndSleep(void)
+{
+    reg_POWCNT2 |= 1; // enable speakers
+    PMi_SetControl(1);
+    SVC_SoundBiasSet(0x100);
+    OS_SpinWait(0x7AB80); // what does this wait for and how long does it wait?
```
Comment (red031000) [inline arm7/lib/src/SND_global.c:48]: it waits for 0x7AB80
:Kappa:

File: arm7/lib/src/SND_seq.c [hunk 22]
```diff
@@ -0,0 +1,1323 @@
...
+    ply->flags.active = TRUE;
+    ply->flags.prepared = FALSE;
+
+    if (SNDi_SharedWork)
+    {
+        SNDi_SharedWork->playerStatus |= 1 << player;
+    }
+}
+
+void SND_StartPreparedSeq(int player)
```
Comment (red031000) [inline arm7/lib/src/SND_seq.c:164]: s32/u32?

File: arm7/lib/src/SND_seq.c [hunk 23]
```diff
@@ -0,0 +1,1323 @@
...
+        switch (inst.type)
+        {
+            case SND_INST_PCM:
+            case SND_INST_DIRECTPCM:
+                // all channels support PCM
+                allowedChannels = 0xFFFF;
+                break;
+            case SND_INST_PSG:
+                // only channels 8, 9, 10, 11, 12, 13 support PSG
+                allowedChannels = 0x3F00;
```
Comment (red031000) [inline arm7/lib/src/SND_seq.c:777]: this feels like there should be constants here

File: arm7/lib/src/SND_seq.c [hunk 24]
```diff
@@ -0,0 +1,1323 @@
...
+
+    if (track->envSustain != 0xFF)
+        SND_SetExChannelSustain(chn, track->envSustain);
+
+    if (track->envRelease != 0xFF)
+        SND_SetExChannelRelease(chn, track->envRelease);
+
+    chn->sweepPitch = track->sweepPitch;
+    if (track->flags.portamento)
+        chn->sweepPitch += (short)((track->portamentoKey - midiKey) << 6);
```
Comment (red031000) [inline arm7/lib/src/SND_seq.c:823]: s16? u16?

File: arm7/lib/src/SND_seq.c [hunk 25]
```diff
@@ -0,0 +1,1323 @@
...
+        chn->sweepLength = swp;
+    }
+
+    chn->sweepCounter = 0;
+}
+
+static int TrackStepTicks(
+    struct SNDTrack *track, struct SNDPlayer *player, int trackIdx, u32 playNotes)
+{
+    (void)trackIdx;
```
Comment (red031000) [inline arm7/lib/src/SND_seq.c:844]: pragma unused?

File: arm7/lib/src/SND_seq.c [hunk 26]
```diff
@@ -0,0 +1,1323 @@
...
+    SeqCacheFetch(track->cur);
+
+    while (track->wait == 0 && !track->flags.noteFinishWait)
+    {
+        specialValueType = FALSE;
+        runCmd = TRUE;
+
+        cmd = TrackReadU8(track);
+
+        if (cmd == 0xA2)
```
Comment (red031000) [inline arm7/lib/src/SND_seq.c:888]: should cmds have constants? I imagine the list may get kinda long if they do

File: arm7/lib/src/SND_util.c [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+    if (shift <= 0)
+    {
+        shift = -shift;
+        result >>= shift;
+    }
+    else if (shift < 32)
+    {
+        // what ???
+        u64 tmp = result & ~0uLL << (32 - shift);
```
Comment (red031000) [inline arm7/lib/src/SND_util.c:40]: wut

**Review by red031000 — CHANGES_REQUESTED** [review]
ints to s32/u32 would be preferable, since this is NitroSDK, but that's not blocking. The constants are

File: arm7/lib/src/SND_seq.c (hunk 26 above)
Comment (ipatix) [reply arm7/lib/src/SND_seq.c:888]: I've thought about it and then I kinda realized those constants would be used in that single case only. Maybe still a good idea, but I for me found it easier to actually see which byte represents which functionality (though adding a comment to summarize functionality might be a good idea).

File: arm7/lib/src/SND_util.c (hunk 27 above)
Comment (ipatix) [reply arm7/lib/src/SND_util.c:40]: Yeah, I think that was one of the things that I genuinely did not understand when I was decompiling it.

File: arm7/lib/src/SND_seq.c (hunk 24 above)
Comment (ipatix) [reply arm7/lib/src/SND_seq.c:823]: Oops, will fix.

File: arm7/lib/src/SND_seq.c (hunk 25 above)
Comment (ipatix) [reply arm7/lib/src/SND_seq.c:844]: What do you mean?

Comment (red031000) [reply arm7/lib/src/SND_seq.c:844]: I mean, rather than doing
(void)trackIdx
can you do
#pragma unused (trackIdx)

Comment (ipatix) [reply arm7/lib/src/SND_seq.c:844]: I would disagree because doing `(void)var` is more standard compliant and it's exactly how you are supposed to do it.

Comment (red031000) [reply arm7/lib/src/SND_seq.c:844]: I guess, though in other places in the code we do use #pragma unused

File: arm7/lib/src/SND_channel.c (hunk 14 above)
Comment (ipatix) [reply arm7/lib/src/SND_channel.c:170]: FFF5 is a mask where bit 1 and bit 3 are cleared (see comment)

File: arm7/lib/src/SND_seq.c (hunk 25 above)
Comment (ipatix) [reply arm7/lib/src/SND_seq.c:844]: Then I'd suggest to change the code to adapt the standard compliant way.

File: arm7/lib/src/SND_seq.c (hunk 24 above)
Comment (ProjectRevoTPP) [reply arm7/lib/src/SND_seq.c:823]: was this fixed yet?

File: arm7/lib/src/SND_seq.c (hunk 23 above)
Comment (ipatix) [reply arm7/lib/src/SND_seq.c:778]: Not really, this is just a bit mask with the bits set which are mentioned in the comment. No further meaning behind that value.

File: arm7/lib/include/PM.h (hunk 1 above)
Comment (ipatix) [reply arm7/lib/include/PM.h]: Actually, I've realized both are probably wrong. I replaced it with u8.

**Review by red031000 — APPROVED** [review]

**Pull request merged**